cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -153,32 +176,53 @@ public String get(String key, String nodeName) {
     return null;
   }
 
-  @Nullable
   @Override
-  public String get(String key) {
-    List<NodeValueEntry> nodeValueList = tokenValueMap.get(key);
-    if (nodeValueList == null) {
-      return null;
+  public List<NodeValueEntry> getAll(String key) {
+    return getAll(key, Scope.USER);
+  }
+
+  @Override
+  public List<NodeValueEntry> getAll(String key, Scope scope) {
+    if (tokenValueMap.get(scope).containsKey(key)) {
+      return ImmutableList.copyOf(tokenValueMap.get(scope).get(key));
     }
+    return ImmutableList.of();
+  }
 
-    if (nodeValueList.isEmpty()) {
-      // List of NodeValueEntry cannot be empty if the key is added in the WorkflowToken as
-      // when we add key, we also add single NodeValueEntry.
-      throw new IllegalStateException(String.format("List of NodeValueEntry for the key %s cannot be empty", key));
+  @Override
+  public Map<String, Value> getAllFromNode(String nodeName) {
+    return getAllFromNode(nodeName, Scope.USER);
+  }
+
+  @Override
+  public Map<String, Value> getAllFromNode(String nodeName, Scope scope) {
+    Map<String, Value> tokenValuesFromNode = new HashMap<>();
+    for (Map.Entry<String, List<NodeValueEntry>> entry : tokenValueMap.get(scope).entrySet()) {
+
+      List<NodeValueEntry> nodeValueEntryList = entry.getValue();
+      for (NodeValueEntry nodeValueEntry : nodeValueEntryList) {
+        if (nodeValueEntry.getNodeName().equals(nodeName)) {
+          tokenValuesFromNode.put(entry.getKey(), nodeValueEntry.getValue());
+          break;
+        }
+      }
     }
-    return nodeValueList.get(nodeValueList.size() - 1).getValue();
+    return tokenValuesFromNode;
```

Comment:
So this method returns a new, mutable map. Other methods in this class returns immutable collection/map. It's better be consistent.